bpo-25658: Implement PEP 539 for Thread Specific Storage (TSS) API

ma8ma opened the pull request:
~~Do not merge until approved the PEP final!~~
Update (2017-09-09): PEP 539 has been accepted.

## Thread Specific Storage (TSS) API

Implement [PEP 539][] for Thread Specific Stroage (TSS) API: it is a new Thread Local Storage (TLS) API to CPython which would supersede use of the existing TLS API within the CPython interpreter, while deprecating the existing API.

### Highlights of changes

- Add Thread Specific Storage (TSS) API
- Add test for transitions of thread key state
- Mark deprecation to Thread Local Storage (TLS) API
- Replace codes that used TLS API with TSS API
- Add the TSS API in the Windows stable ABI (https://github.com/python/cpython/pull/1362#issuecomment-299707811)
- ~~[TEMP] Add an example function to ``xxlimited`` module~~

[PEP 539]: https://www.python.org/dev/peps/pep-0539/
<!-- issue-number: bpo-25658 -->
https://bugs.python.org/issue25658
<!-- /issue-number -->

ma8ma committed:
Implement PEP 539 Thread Specific Storage (TSS) API

See PEP 539 for details.

highlights of changes:

- Add Thread Specific Storage (TSS) API
- Add test for transitions of the thread key state.
- Mark deprecation to Thread Local Storage (TLS) API
- Replace codes that used TLS API with TSS API

AraHaan commented:
Uh are you rebasing? I made that mistake before. The latest dev guide says to not rebase but include all commits which would then be merged as a single commit.

ma8ma commented:
@AraHaan Yes, PR is single commit. I confirmed AppVeyor build failed, and retried push in a minute. But it has failed again, I need help for Windows platform...

AraHaan commented:
It seems to be something with the solution or project settings or even AppVeyor's setting. Will investigate more when I can use the internet on my Inspiron 1545 that has Windows 7 Ultimate SP1 x64 and VS2015 installed. It seems that in the project settings is set to use the Windows sdk for 8.1 instead of for 10. Also when targeting for 10.0 I recommend staticallylinking the ucrt for the versions of Windows that can't install it (Vista). Also it allows flexibility and usability on fresh installs of 7 and Vista do that is another reason why I recommend it.

ncoghlan commented on Include/pythread.h:
```diff
@@ -77,15 +79,67 @@ PyAPI_FUNC(int) PyThread_set_stacksize(size_t);
+
+/* Py_tss_t is opaque type and you *must not* directly read and write.
+   When you'd check whether the key is created, use PyThread_tss_is_created.
+*/
```
This comment should explain that the type layout is public (despite being nominally opaque) to allow for static declarations in API clients.

It should be made fully opaque when `Py_LIMITED_API` is defined, though: https://docs.python.org/3/c-api/stable.html

ncoghlan commented on Include/pythread.h:
```diff
@@ -77,15 +79,67 @@ PyAPI_FUNC(int) PyThread_set_stacksize(size_t);
+PyThread_tss_is_created(Py_tss_t key)
+{
+    return key._is_initialized;
+}
```
To allow for use in the `Py_LIMITED_API` case, the public APIs that currently accept `Py_tss_t` by value will need to accept it as a pointer instead.

The alternative would be to make the `Py_tss_t` struct layout part of the stable ABI, but I'm reluctant to do that due to the `NATIVE_TLS_KEY_T` field: it would keep us from switching from `int` to a different native type for platforms that don't currently have a native implementation, but gain one in the future.

Following that chain of thought further, I guess fully supporting Py_LIMITED_API would require an extra pair of memory management functions (`PyThread_tss_alloc` and `PyThread_tss_free`), as with the struct definition hidden, API clients won't be able to use static variables.

embray replied:
I'm not super-familiar with the constraints regarding `Py_LIMITED_API`.  Addressing this will require an update to the PEP as well.

I wouldn't mind if the functions in question were changed to take a pointer argument instead of passing the struct by value just for consistency's sake if nothing else.

ma8ma replied:
Okay, I start work to modify to full ``Py_LIMITED_API`` on the weekend.

embray replied:
With a `PyThread_tss_alloc`, would the need for `Py_tss_NEEDS_INIT` to be public go away?  After all, if one used `PyThread_tss_alloc` to allocate memory for a struct, what else would they even initialize it with?  It would remove an extra step to just do that as part of `PyThread_tss_alloc` and remove that implementation detail.  

It should just be clear that just because a TSS key has been *allocated* doesn't mean it's been *created*.  Also, `PyThread_tss_is_created` is still available to test whether or not a key has been created.

ncoghlan replied:
static storage allocation would still be supported for the normal (non-limited API) case, so the public `Py_tss_NEEDS_INIT` initializer would still be needed.

ncoghlan commented on Python/thread.c:
```diff
@@ -247,47 +248,114 @@ find_key(int set_value, int key, void *value)
+
+/* Thread Local Storage (TLS) API, DEPRECATED since Python 3.7
+
+   In the own implementation, TLS API has been changed to call TSS API.
```
For the second sentence here, I'd suggest "While the native thread-local storage implementations have been retained, this fallback implementation is now just a compatibility wrapper around the TSS API."

ncoghlan commented on Python/thread_foobar.h:
```diff
@@ -123,10 +126,74 @@ PyThread_delete_key_value(int key)
+
+void
+PyThread_tss_delete_value(Py_tss_t key)
+{
```
Does PEP 7 express  a preference regarding empty function definitions? If it doesn't we should at least make sure the TLS/TSS code is self-consistent (one of the later changes has the opening and closing brace on the same line).

embray replied:
Not that I can tell, in which case the most consistent style would be that demonstrated here.

ma8ma replied:
Sure, I'm going to make formatting even. By the way, would I create an issue for removing ``Python/thread_foobar.h``? Although I'm not sure history of thread_foobar.h, it seems to me that is dead code at present.

ncoghlan replied:
@ma8ma Aye, a separate issue and python-dev thread for that sounds sensible to me. The question to the mailing list can just be something simple like "This looks like dead code and we want to delete it, does anyone have a reason we should keep it?"

embray replied:
I had assumed it was a dummy implementation for testing / demonstration purposes.  It might be an easier way to understand the requirements of the threading API without being bogged down with the details of specific implementations.  That said, I don't know how useful that is in practice. I don't think I ever looked at it, opting instead to look at the pthread implementation since pthreads are already familiar...

ma8ma replied:
Well, in fact, I think it makes the loading to maintainers. I posted to python-dev.

ncoghlan reviewed (changes requested):
This looks excellent to me as a solution for the standard ABI, but I realised in reviewing it that the stable ABI poses a potential problem: we don't want to expose the `Py_tss_t` definition in that case, but making the API usable with an opaque type definition would require a few changes (details inline).

ncoghlan commented on Include/pythread.h:
```diff
@@ -77,15 +79,67 @@ PyAPI_FUNC(int) PyThread_set_stacksize(size_t);
+*/
+
+#if defined(_POSIX_THREADS)
+#   define NATIVE_TLS_KEY_T     pthread_key_t
```
For consistency, this should be `NATIVE_TSS_KEY_T`

ma8ma committed:
bpo-25658: Modify comment about wrapper function

https://github.com/python/cpython/pull/1362#discussion_r114079176

ma8ma committed:
bpo-25658: Rename native TSS key type macro

https://github.com/python/cpython/pull/1362#discussion_r114473591

ma8ma committed:
bpo-25658: Change API interfaces to provide even in the limited API

https://github.com/python/cpython/pull/1362#discussion_r114079006

ma8ma commented:
In Windows, the module-definition (.def) file for limited API (python3.dll) doesn't have the TSS API (also the TLS API), therefore, ``xxlimited`` module build occurs [link error][appveyor]. ~~Would I append TSS API to [python3.def][python3_def] to update stable ABI?~~

**Edit (2017-05-13):** Add WIP commit for Windows ABI
**Edit (2017-08-31):** ~~As far as someone doesn't show a necessary reason, I'd revert WIP commit that adding TSS API into Windows stable ABI. (i.e. back to same support level as the existing TLS API)~~
**Update (2017-09-02):** ~~Revert a temporary commit which supports TSS API in Windows stable ABI, because the existing TLS API is not provided in Windows stable ABI. TSS API support goes to maintain the status quo.~~
**Update (2017-09-18):** Add the commit again (see https://github.com/python/cpython/pull/1362#issuecomment-330107896).

[PCbuild description](https://github.com/python/cpython/tree/master/PCbuild)

[appveyor]: https://ci.appveyor.com/project/python/cpython/build/3.7.0a0.1850
[python3_def]: https://github.com/python/cpython/blob/master/PC/python3.def

ma8ma committed:
bpo-25658: Add TSS API to Windows stable ABI

https://ci.appveyor.com/project/python/cpython/build/3.7.0a0.1850

ma8ma committed:
bpo-25658: Add the TSS API to Windows stable ABI

ma8ma committed:
bpo-25658: Add NEWS and What's New entries

ma8ma committed:
bpo-25658: Add an example function to xxlimited module

Show a concrete example using TSS API under the limited API.

ma8ma committed:
bpo-25658: Add braces to follow updated PEP 7 format

https://github.com/python/peps/issues/279

ma8ma committed:
bpo-25658: Format empty function definitions

https://github.com/python/cpython/pull/1362#discussion_r114079269

ma8ma committed:
bpo-25658: Clarify comments for Py_tss_t type explanations

https://github.com/python/cpython/pull/1362#discussion_r114078743

ma8ma committed:
Merge commit 'master' into pep539-tss-api

Resolve conflicts:
fdaeea6 bpo-30279: Remove unused Python/thread_foobar.h (#1473)

ma8ma committed:
Merge commit 'master' into pep539-tss-api

Resolve conflcts:
ab4413a bpo-30039: Don't run signal handlers while resuming a yield from stack (#1081)

ma8ma committed:
Merge commit 'master' into pep539-tss-api

Resolve conflicts:
346cbd3 bpo-16500: Allow registering at-fork handlers (#1715)

ma8ma committed:
bpo-25658: Rename ReInitTLS function in Modules/posixmodule.c

The code has been moved from Modules/signalmodule.c. (346cbd3)

ma8ma committed:
Merge commit 'master' into pep539-tss-api

Resolve conflicts:
f7ecfac Doc nits for bpo-16500 (#1841)

ma8ma committed:
Merge commit 'master' into pep539-tss-api

Resolve conflicts:
3b5cf85 bpo-30524: Write unit tests for FASTCALL (#2022)

ma8ma committed:
Merge branch 'master' into pep539-tss-api

ma8ma committed:
bpo-25658: Add news using blurb

New workflow was announced on Python-Dev.

ma8ma committed:
Merge branch 'master' into pep539-tss-api

Resolve conflicts:
aa0aa04 bpo-30832: Remove own implementation for thread-local storage (#2537)

ma8ma committed:
bpo-25658: Omit PyThread_ReInitTSS function

After aa0aa04, the own implementation for TLS is gone and
PyThread_ReInitTSS function has made to be no-op for all supported
platforms (Windows and pthreads).

ma8ma committed:
Merge commit 'master' into pep539-tss-api

Resolve conflicts:
64fa449 bpo-30866: Add _testcapi.stack_pointer() (#2601)

ma8ma committed:
bpo-25658: Set error directive on pythread.h to require native thread

ma8ma committed:
bpo-25658: Omit PyThread_tss_delete_value function

After aa0aa04, the own implementation for TLS is gone and
PyThread_tss_delete_value function just does to empty the storage
(i.e. "PyThread_tss_delete_value(key)" has made to be equal to
"PyThread_tss_set(key, NULL)").

ma8ma committed:
Merge branch 'master' into pep539-tss-api

ma8ma committed:
bpo-25658: Reorder the test function

ma8ma committed:
bpo-25658: Update comments

ma8ma committed:
bpo-25658: Enclose Py_tss_t definition with the Py_LIMITED_API macro

Make more clearly the switch of show/hide implementation detail by
the macro definition

ma8ma committed:
Merge branch 'master' into pep539-tss-api

ma8ma committed:
bpo-25658: follow the draft specification

eamanu reviewed (approved):
All changes agrees with PEP7

zooba commented on Include/pythread.h:
```diff
@@ -77,15 +83,71 @@ PyAPI_FUNC(int) PyThread_set_stacksize(size_t);
+#undef NATIVE_TSS_KEY_T
+
+/* When static allocation, you must initialize with Py_tss_NEEDS_INIT. */
+#define Py_tss_NEEDS_INIT   {._is_initialized = false}
```
Please don't use designated initializers in header files - they prevent people using Python in C++.

ma8ma replied:
Yeah, I'll do so.

embray replied:
I'm not exactly sure what the alternative would be...?

ncoghlan replied:
Using a partial positional initializer: `#define Py_tss_NEEDS_INIT   {false,}`

The struct is simple enough that we can live without the extra clarity of repeating the field name for the benefit of better C++ compatibility.

zooba commented on Include/pythread.h:
```diff
@@ -2,6 +2,12 @@
+#   error "Require native thread feature. See https://bugs.python.org/issue30832"
+#endif
+
+#include <stdbool.h>  /* necessary for TSS key */
```
Why is this necessary? Just use an `int` for the initialized flag and avoid the extra dependency - every extra header can be a big deal for non-desktop compilers.

ma8ma replied:
Aye, I'd change PyThread_tss_is_created declaration. First of all, I report your proposal to python-dev and python/peps#382.

embray replied:
I prefer `bool` myself, but Steve has a point here. There's not much advantage if it means bringing in a header that wouldn't otherwise be necessary, and isn't (currently) used anywhere else in Python's headers.  If we change this it will require a minor update to the PEP.

zooba commented on Include/pythread.h:
```diff
@@ -77,15 +83,71 @@ PyAPI_FUNC(int) PyThread_set_stacksize(size_t);
+*/
+typedef struct _Py_tss_t Py_tss_t;
+
+#ifndef Py_LIMITED_API
```
This should do a version check on the macro, since these APIs are not in the limited API prior to 3.7.

ma8ma replied:
That's true, I'd enclose TSS API declaration.

zooba reviewed (commented):
I'd very much like to see the header files cleaned up to reduce potential for compatibility issues, but otherwise the implementation looks fine to me.

ma8ma committed:
bpo-25658: change the declaration of PyThread_tss_is_created

Use an int for the initialized flag and avoid the extra dependency.
https://github.com/python/cpython/pull/1362#pullrequestreview-59884901

ma8ma committed:
bpo-25658: use a partial positional initializer for Py_tss_NEEDS_INIT

Using designated initializers in header files prevents us using Python
in C++.
https://github.com/python/cpython/pull/1362#pullrequestreview-59884205

ma8ma committed:
bpo-25658: add the version check for the limited API

TSS API is not in the limited API prior to CPython 3.7.
https://github.com/python/cpython/pull/1362#pullrequestreview-59885908

ma8ma committed:
Revert "bpo-25658: Add an example function to xxlimited module"

This reverts commit d25f441e6ac8eef69e93881c03044139213abf6b.
Since the specification is almost fixed, revert a temporary commit
which builds xxlimited extension module to use the TSS API in the
limited API.

ma8ma committed:
Revert "bpo-25658: Add TSS API to Windows stable ABI"

This reverts commit 594ec29a07794f7f1748c5923362ad1f77e14dc5.
Revert a temporary commit which supports TSS API in Windows stable
ABI, because the existing TLS API is not provided in Windows stable
ABI. TSS API support goes to maintain the status quo.

ma8ma committed:
bpo-25658: fix an assignment to the _is_initialized field in PyThread_tss_alloc

ma8ma committed:
Merge commit 'master' into pep539-tss-api

Resolve conflicts:
a6a4dc8 bpo-31370: Remove support for threads-less builds (#3385)

ma8ma committed:
bpo-25658: remove WITH_THREAD guard

ma8ma commented:
As above, PEP 539 has been accepted, ~~but [WIP] has been labeled yet~~.
By the way, I wrote a commit message that summarizes this PR, because this PR has too much commits. The changes won't be enumerated one by one.
```
bpo-25658: Implement PEP 539 for Thread Specific Storage (TSS) API

Add a new Thread Local Storage (TLS) API to CPython which would
supersede use of the existing TLS API within the CPython interpreter,
while deprecating the existing API.  The change is motivated
primarily by the fact that the old API uses int to represent TLS keys
across all platforms, which is neither POSIX-compliant, nor portable
in any practical sense.  This change covers not only the new API
implementation features, but also the client code updates needed to
replace the existing TLS API with the new TSS API.
```

ma8ma committed:
Merge branch 'master' into pep539-tss-api

Resolve conflicts:
2ebc5ce bpo-30860: Consolidate stateful runtime globals. (#3397)

ma8ma committed:
bpo-25658: add contiributor names to news and whatsnew

ma8ma committed:
bpo-25658: update compile error message in thread-less builds

Remove tracemalloc's message that has made no longer sensible for
since all platforms require native threads.

ma8ma committed:
bpo-25658: create PEP 539 topic for New Features in whatsnew

ma8ma committed:
bpo-25658: put header's error directive into Py_LIMITED_API guard

Code smell breaking backward compatibility in the limited API.

ma8ma commented on Python/pystate.c:
```diff
@@ -46,7 +46,12 @@ _PyRuntimeState_Init(_PyRuntimeState *runtime)
+    {
+        Py_tss_t initial = Py_tss_NEEDS_INIT;
+        runtime->gilstate.autoTSSkey = initial;
+    }
```
I have another idea instead of assignment by instant variable; provide a macro or inline function which ready the key (put into `Include/pythread.h`?). Edit: The name `_READY` is different behavior to `PyUnicode_READY` that returns a status code. `Py_tss_INITIALIZE (_PyThread_tss_initialize)` is more make sense?
```c
#ifndef Py_LIMITED_API
/* You *must not* use except for the case of initializing the key
   in the startup sequence! */
#define Py_tss_READY(key) do { (key)._is_initialized = 0; } while(0)
static inline void
PyThread_tss_ready(Py_tss_t *key) {
    key->_is_initialized = 0; 
}
#endif
```
In addition, calling `PyThread_tss_delete` gives to reset the key's initialization state, but it will shake the create/use/delete cycle. I don't like it.

ncoghlan replied:
Since this is an internal use, I think it would be OK to just write `runtime->gilstate.autoTLSkey._is_initialized = 0`.

However, I don't consider this a merge blocker either way - the `Py_tss_NEEDS_INIT` version also works, and we can make "Provide a nicer runtime TSS structpre-initialization API" a separate RFE.

ma8ma replied:
I'd like to maintain the status quo to indicate that `Py_tss_t` is an opaque type, this is the one of the few chance.  In fact, except for testing, there is only another example `Module/_tracemalloc.c` in CPython repo. This is also the reason that first I chose the assignment by struct, but not the feature of initializing struct field (I'm not sure it will be required...).

ma8ma committed:
bpo-25658: show an API difference on whatsnew by initialization codes

ncoghlan commented on Include/pythread.h:
```diff
@@ -77,15 +77,76 @@ PyAPI_FUNC(int) PyThread_set_stacksize(size_t);
+    */
+#   define NATIVE_TSS_KEY_T     unsigned long
+#else
+#   error "Require native threads. See https://bugs.python.org/issue30832"
```
This ref can also be updated to 31370

ncoghlan commented on Python/thread.c:
```diff
@@ -139,14 +139,46 @@ happen to be PyObject*, these functions don't do refcount operations on
 
+/* Thread Specific Storage (TSS) API
+
+   Implementation part of common
```
Suggested rephrasing for this comment: "Cross-platform components of TSS API implementation"

ncoghlan commented on Python/thread_nt.h:
```diff
@@ -349,10 +349,12 @@ _pythread_nt_set_stacksize(size_t size)
 
 
+/* Thread Local Storage (TLS) API, DEPRECATED since Python 3.7 */
+
```
May be worth mentioning PEP 539 specifically here.

ncoghlan commented on Python/thread_nt.h:
```diff
@@ -399,9 +397,74 @@ PyThread_delete_key_value(int key)
+
+/* Thread Specific Storage (TSS) API
+
+   Implementation part of platform-specific
```
Rephrasing inline with the above suggestion: "Platform-specific components of TSS API implementation."

ma8ma replied:
Sure, I fix comments in one go.

ncoghlan reviewed (approved):
Functionally, this is looking good to me now.

@ma8ma Would you be OK with my just going ahead and making the comment fixes I suggest? I'd also like to reword the What's New additions a bit, and add a new section to https://docs.python.org/3/c-api/init.html for Thread Specific Storage (since this is officially a public API now).

ma8ma committed:
Merge branch 'master' into pep539-tss-api

ma8ma committed:
bpo-25658: fix an issue number in the error directive

https://github.com/python/cpython/pull/1362#discussion_r139301168

ma8ma committed:
bpo-25658: fix comments

https://github.com/python/cpython/pull/1362#discussion_r139301183
https://github.com/python/cpython/pull/1362#discussion_r139301195
https://github.com/python/cpython/pull/1362#discussion_r139301215

ma8ma committed:
bpo-25658: update a comment the pthreads for TLS implementation

ma8ma commented:
@ncoghlan Okay, I updated comments!

> add a new section to https://docs.python.org/3/c-api/init.html for Thread Specific Storage (...).

I'll try that too. (well, take time...)

> since this is officially a public API now

Does that mean the TSS API is part of Windows stable ABI? (i.e. commit 594ec29 again except for PyThread_tss_delete_value)

ncoghlan commented:
@ma8ma Yes, the parts protected by `#ifdef Py_LIMITED_API+0 >= 0x03070000` would all qualify for inclusion in the Windows stable ABI.

However, even without that, things can be part of the public C API without being part of the stable ABI - it just means that they remain source-code compatible between releases, but not necessarily binary compatible. The `Py_tss_t` struct is an example of that here - it's not private (although the individual field names are), so if you're willing to recompile for every new CPython feature release, you can use it as a static type, rather than dynamically allocating it.

ma8ma commented:
@ncoghlan Thank you for the explanation. I don't think there is any problem, so I add the TSS API in the Windows stable ABI.

ma8ma committed:
Merge branch 'master' into pep539-tss-api

ma8ma committed:
bpo-25658: remove API description on the code

ma8ma committed:
bpo-25658: add API document

- overview
- the existing TLS API (without descripiton)
- the new TSS API
  - type and macro
  - dynamic allocation
  - methods

ma8ma committed:
bpo-25658: modify whatsnew document

ma8ma committed:
bpo-25658: modify header comments

Several comments is no longer necessary by API document.

ma8ma committed:
bpo-25658: update order of the declaration

New order is along API document.

ma8ma commented:
I updated this PR to add API document. Sources of the API description are:

- [PEP 539](https://www.python.org/dev/peps/pep-0539/)
- comments on the code (5073d66)

Documents rendered by github:

- [c-api/init.rst](https://github.com/ma8ma/cpython/blob/pep539-tss-api/Doc/c-api/init.rst#thread-local-storage-support)
- [whatsnew/3.7.rst](https://github.com/ma8ma/cpython/blob/pep539-tss-api/Doc/whatsnew/3.7.rst#pep-539-a-new-c-api-for-thread-local-storage-in-cpython)

ncoghlan reviewed (changes requested):
This is looking very good to me. I'm going to go through and make some edits to the docs updates, and add in the required limited API version check, but after that I think this should be ready to merge.

ncoghlan commented:
Oops, the required macro version check is there, it just wasn't visible in the subset of changes I was looking at - so I'll just make the docs tweaks, and then this will be ready to merge.

embray commented:
Great! I was just about to say, since `--without-threads` was removed it's impossible to compile on Cygwin without this now.  Is there anything else I can do to help or is this otherwise ready to go?

ncoghlan committed:
Adjust some details of documentation wording

ncoghlan commented:
@embray I think it's good to go once the latest CI run finishes, but another set of eyes looking over the changes wouldn't hurt (if it's just docs comments though, then it would probably make sense to defer those to a new PR so we can restore Cygwin compatibility in the meantime)

ncoghlan commented:
And merged - thank you @ma8ma for the PR!

ma8ma commented:
@embray @ncoghlan Thank you for your advice too! :smile:

shaorongchen88 reviewed (commented):
<deleted as the comment seemed like spam>

qwhz reviewed (commented):
- [ ] [](url)
